src/lsm/compaction.zig

```diff
@@ -281,1843 +302,1623 @@ pub fn CompactionType(
+                        compaction.level_a_value_block_next = 0;
+                    }
+                },
+                .level_b => {
+                    compaction.level_b_value_block_next += 1;
+                    if (compaction.level_b_value_block_next ==
+                        index_schema.data_blocks_used(index_block.ptr))
+                    {
+                        value_block.last_block_in_the_table = true;
+                        const pop = compaction.level_b_index_block.pop().?;
+                        assert(pop.ptr == index_block.ptr);
+                        assert(index_block.stage == .read_index_block_done);
+
+                        compaction.read_value_block_release_table(index_block.ptr);
+                        index_block.stage = .free;
+                        compaction.pool.?.block_release(index_block);
+                        compaction.level_b_value_block_next = 0;
+                    }
+                },
             }
-
-            if (finished_value_block) return .finished_value_block;
-            return .unfinished_value_block;
         }
 
         // TODO: Support for LSM snapshots would require us to only remove blocks
         // that are invisible.
-        fn release_table_blocks(compaction: *Compaction, index_block: BlockPtrConst) void {
-            // Release the table's block addresses in the Grid as it will be made invisible.
-            // This is safe; compaction.index_block_b holds a copy of the index block for a
-            // table in Level B. Additionally, compaction.index_block_a holds
-            // a copy of the index block for the Level A table being compacted.
-            log.debug("release_table_blocks({s})", .{compaction.tree_config.name});
-
-            const grid = compaction.grid;
+        fn read_value_block_release_table(
+            compaction: *Compaction,
+            index_block: BlockPtrConst,
+        ) void {
             const index_schema = schema.TableIndex.from(index_block);
-            for (index_schema.data_addresses_used(index_block)) |address| grid.release(address);
-            grid.release(Table.block_address(index_block));
+            const index_block_address = Table.block_address(index_block);
+            const data_block_addresses = index_schema.data_addresses_used(index_block);
+            // Tables are release when the index block is no longer needed. Given that the same
+            // index block can get re-read across the bar, the same table can be released twice.
+            if (compaction.grid.free_set.is_released(index_block_address)) {
+                for (data_block_addresses) |address| {
+                    assert(compaction.grid.free_set.is_released(address));
+                }
+            } else {
+                for (data_block_addresses) |address| {
+                    compaction.grid.free_set.release(address);
+                }
+                compaction.grid.free_set.release(index_block_address);
```

Comment:
>Ah, so we use free_set.release() instead of grid.release() b/c we don't want to demote the blocks from the cache, as they may be re-read again during the next beat(s).

No, that is a genuine refactoring bug! I didn't intend to fix this.

That being said, this feels a bit backwards --- it seems that, rather than demoting tables after the fact, the grid should perhaps read them without bumping the cache in the first place? 

>Let's assert in bar_commence() for each (level A and B) index block address:

Ah, this is great!